src/dso/dso_transport.cpp

```diff
@@ -0,0 +1,697 @@
+
+    if (it != mMap.end())
+    {
+        ret = it->second.get();
+    }
+
+    return ret;
+}
+
+DsoAgent::DsoConnection *DsoAgent::FindOrCreateConnection(otPlatDsoConnection *aConnection)
+{
+    auto &ret = mMap[aConnection];
+
+    if (!ret)
+    {
+        ret = MakeUnique<DsoConnection>(this, aConnection);
+    }
+
+    return ret.get();
+}
+
+DsoAgent::DsoConnection *DsoAgent::FindOrCreateConnection(otPlatDsoConnection *aConnection, mbedtls_net_context aCtx)
+{
+    auto &ret = mMap[aConnection];
+
+    if (!ret)
+    {
+        ret = MakeUnique<DsoConnection>(this, aConnection, aCtx);
+    }
+
+    return ret.get();
+}
+
+void DsoAgent::Enable(otInstance *aInstance)
+{
+    OTBR_UNUSED_VARIABLE(aInstance);
+
+    assert(aInstance == mInstance);
+
+    constexpr int kOne = 1;
+    sockaddr_in6  sockAddr;
+
+    VerifyOrExit(!mListeningEnabled);
+
+    mListeningCtx.fd = socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP);
+
+    VerifyOrExit(
+        !setsockopt(mListeningCtx.fd, SOL_SOCKET, SO_BINDTODEVICE, mInfraNetIfName.c_str(), mInfraNetIfName.size()));
+    VerifyOrExit(!setsockopt(mListeningCtx.fd, SOL_SOCKET, SO_REUSEADDR, (const uint8_t *)&kOne, sizeof(kOne)));
+    VerifyOrExit(!setsockopt(mListeningCtx.fd, SOL_SOCKET, SO_REUSEPORT, (const uint8_t *)&kOne, sizeof(kOne)));
+
+    sockAddr.sin6_family = AF_INET6;
+    sockAddr.sin6_addr   = in6addr_any;
+    sockAddr.sin6_port   = htons(kListeningPort);
+    VerifyOrExit(!bind(mListeningCtx.fd, reinterpret_cast<sockaddr *>(&sockAddr), sizeof(sockAddr)));
+    VerifyOrExit(!mbedtls_net_set_nonblock(&mListeningCtx), otbrLogWarning("Failed to set non-blocking: %d", errno));
+    VerifyOrExit(!listen(mListeningCtx.fd, kMaxQueuedConnections));
+
+    mListeningEnabled = true;
+
+    otbrLogInfo("DSO socket starts listening");
+
+exit:
+    return;
+}
+
+void DsoAgent::Disable(otInstance *aInstance)
+{
+    OTBR_UNUSED_VARIABLE(aInstance);
+
+    assert(aInstance == mInstance);
+
+    VerifyOrExit(mListeningEnabled);
+
+    mbedtls_net_close(&mListeningCtx);
+    mbedtls_net_free(&mListeningCtx);
+    mMap.clear();
+    mListeningEnabled = false;
+
+exit:
+    return;
+}
+
+void DsoAgent::SetEnabled(otInstance *aInstance, bool aEnabled)
+{
+    if (aEnabled)
+    {
+        Enable(aInstance);
+    }
+    else
+    {
+        Disable(aInstance);
+    }
+}
+
+void DsoAgent::RemoveConnection(otPlatDsoConnection *aConnection)
+{
+    mMap.erase(aConnection);
+}
+
+void DsoAgent::SetHandlers(AcceptHandler       aAcceptHandler,
+                           ConnectedHandler    aConnectedHandler,
+                           DisconnectedHandler aDisconnectedHandler,
+                           ReceiveHandler      aReceiveHandler)
+{
+    mAcceptHandler       = std::move(aAcceptHandler);
+    mConnectedHandler    = std::move(aConnectedHandler);
+    mDisconnectedHandler = std::move(aDisconnectedHandler);
+    mReceiveHandler      = std::move(aReceiveHandler);
+}
```

Comment:
What is the use-case to set different handlers?

Reply:
Please ignore earlier comment. I saw in the comment that this is used for testing. 👍